torchmultimodal/modules/layers/position_embedding.py

```diff
@@ -57,76 +57,66 @@ def __init__(
             >>> out = pos_emb._broadcast(i=0)
             >>> print(out)
-            tensor([[[[0, 0, 0], [0, 0, 0], [0, 0, 0], [0, 0, 0]],
-                    [[0, 0, 1], [0, 0, 1], [0, 0, 1], [0, 0, 1]]]])
+            tensor([[[0, 0, 0], [0, 0, 0], [0, 0, 0], [0, 0, 0]],
+                    [[0, 0, 1], [0, 0, 1], [0, 0, 1], [0, 0, 1]]])
             >>> print(out.shape)
-            (1, 2, 4, 3)
+            (2, 4, 3)
 
         The input is broadcasted along the second dim ``4`` since it's the ``0``-th embedding constructed w.r.t the
         first dim ``2``.
         """
         emb = self.embedding[f"d_{i}"]
-        # (1, 1, ..., 1, self.shape[i], 1, ..., -1)
+        # (1, ..., 1, self.shape[i], 1, ..., embedding_dim)
         emb = emb.view(
-            1,
             *itertools.repeat(1, i),
             self.shape[i],
             *itertools.repeat(1, (self.n_dim - i - 1)),
             -1,
         )
-        # (1, *self.shape, -1)
-        emb = emb.expand(1, *self.shape, -1)
+        # (*self.shape, embedding_dim)
+        emb = emb.expand(*self.shape, -1)
 
         return emb
 
-    def _decode(
-        self, decode_step: int, embeddings: Tensor, x_shape: Tuple[int, ...]
-    ) -> Tensor:
-        """Returns the embedding vector immediately before the decoding location."""
-        decode_idx = self.decode_idxs[decode_step - 1]
-        embeddings = tensor_slice(
-            embeddings,
-            [0, *decode_idx, 0],
-            [x_shape[0], *itertools.repeat(1, self.n_dim), x_shape[-1]],
-        )
-
-        return embeddings
+    def forward(self, position_ids: Tensor) -> Tensor:
+        if torch.max(position_ids) >= len(self.indices) or torch.min(position_ids) < -1:
+            raise IndexError(f"Invalid position ids: {position_ids}")
 
-    def forward(
-        self, x: Optional[Tensor] = None, decode_step: Optional[int] = None
-    ) -> Tensor:
         embeddings = []
         for i in range(self.n_dim):
             emb = self._broadcast(i)
             embeddings.append(emb)
 
-        embeddings = torch.cat(
-            embeddings, dim=-1
-        )  # concatenated embeddings: (1, *(shape), embedding_dim)
-
-        if decode_step is not None:
-            embeddings = self._decode(decode_step, embeddings, tuple(x.shape))
-            # decoded embedding: (1, *repeat(1, len(shape)), embedding_dim)
+        # concatenated embeddings: (*(shape), embedding_dim)
+        embeddings = torch.cat(embeddings, dim=-1)
+        # extract relevant indices for the embeddings and transpose: (len(position_ids), len(shape))
+        indices = [*self.indices[position_ids].transpose(0, 1)]
+        # return the relevant embeddings: (len(position_ids), embedding_dim)
+        # the embeddings are flattened across ``shape`` as the indices are broadcasted
+        embeddings = embeddings[indices]
 
-        return embeddings.flatten(start_dim=1, end_dim=-2)
+        # (batch, len(position_ids), embedding_dim)
```

Comment:
Is this comment referring to the return shape? If so should prepend with a 1, no?